### PR TITLE
[inductor] Add ir.Scan and lower aten.cumsum on CUDA

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #110913
* #110912
* #110911
* #109132
* __->__ #106581

This adds the `ir.Scan` node (currently only supported on CUDA) which re-uses the existing reduction kernel machinery to support different kinds of non-pointwise ops. Just like reductions it supports prologue and epilogue fusions and has both persistent and non-persistent kernel generation.

Currently this doesn't support the equivalent of `Reduction.create_multilayer` and will instead fall back to eager in those cases. This is because splitting into multiple kernel invocations ends up being far slower than cub's single kernel strategy which matches the performance of a copy kernel.

Fixes https://github.com/pytorch/pytorch/issues/93631

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @Xia-Weiwen @ngimel

### CODE REVIEW

File: test/inductor/test_torchinductor.py [hunk 1]
```diff
@@ -996,6 +996,12 @@ def fn(x):
 
         self.common(fn, (torch.ones(32, 32) * 70,))
 
+    def test_cumsum(self):
```
Comment (lezcano) [inline test/inductor/test_torchinductor.py:999]: Can you also add a test for which `persistent_reduction=False`?

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -1359,6 +1362,143 @@ def intermediate_loader_fn(index, reduction_index, loader):
...
+            device=device,
+            dtype=dtype,
+            inner_fn=inner_fn,
+            axis=axis,
+            pointwise_ranges=pointwise_ranges,
+            scan_ranges=scan_ranges,
+            scan_op=scan_op,
+            scan_numel=scan_numel,
+        )
+        if num_splits > 1:
```
Comment (lezcano) [inline torch/_inductor/ir.py:1450]: Add a tiny TODO.

**Comment by peterbell10** [comment]
Okay, I ran the benchmarks to see which ones actually hit cumsum and the only ones are:
- `hf_Longformer`
- `AllenaiLongformerBase`
- `Camembert`
- `M2M100ForConditionalGeneration`
- `RobertaForCausalLM`
- `RobertaForQuestionAnswering`

None of which show a significant change in this benchmark run:
https://hud.pytorch.org/benchmark/compilers?startTime=Wed%2C%2030%20Aug%202023%2015%3A45%3A20%20GMT&stopTime=Wed%2C%2006%20Sep%202023%2015%3A45%3A20%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/peterbell10/590/orig&lCommit=0f1079e2a476b05674a045123529b2dccbbbfe50&rBranch=main&rCommit=06b173780d5c85c8aa35aa615fe2d0377ba61cb4

There are some other differences showing but they appear to just be noise.

**Comment by lezcano** [comment]
A few questions:
- This IR node I guess that it allows for prologue fusion, but does it allow for epilogue fusion? If so, can we fuse anything of the same size as the input after it, as ,for example, a reduction?
- Could you post some benchmark numbers between triton's implementation and core's for the persistent / non persistent case?

I just find so weird that the numbers are the same... But if the efficiency is the same as cub's and we don't fuse anything to it that'd make sense...

**Comment by peterbell10** [comment]
> This IR node I guess that it allows for prologue fusion, but does it allow for epilogue fusion? If so, can we fuse anything of the same size as the input after it, as ,for example, a reduction?

Yes, this supports epilogue fusion and if you happen to fuse a large reduction you can see some real performance wins. e.g.
```
@torch.compile
def fn(x, dim):
    t = x.cumsum(dim)
    return (x / t).sum(dim)

x = torch.testing.make_tensor((1024, 1024), low=0, device="cuda", dtype=torch.float32)
fn(x, -1)
```

Results in a single persistent reduction kernel that doesn't materialize the intermediate cumsum, so saves a lot of bandwidth.
Also, the modification to `end_current_reduction_loop` means that unlike normal reductions, cumulative scans can be scheduled in the same loop as a non-persistent reduction. So, for the example above we can generate:
```
    tmp2 = tl.full([XBLOCK, RBLOCK], 0, tl.float32)
    _tmp10 = tl.full([XBLOCK, RBLOCK], 0, tl.float32)
    for roffset in range(0, rnumel, RBLOCK):
        rindex = roffset + rbase
        rmask = rindex < rnumel
        r1 = rindex
        tmp0 = tl.load(in_ptr0 + (r1 + (4096*x0)), rmask & xmask, other=0)
        tmp1 = tl.broadcast_to(tmp0, [XBLOCK, RBLOCK])
        tmp3 = tl.where(rmask & xmask, tmp1, 0)
        tmp4 = tl.sum(tmp1, 1)[:, None]
        tmp5 = tmp2 + tmp4
        tmp6 = tl.cumsum(tmp3, 1)
        tmp7 = tmp2 + tmp6
        tmp2 = tmp5
        tmp8 = tmp0 / tmp7
        tmp9 = tl.broadcast_to(tmp8, [XBLOCK, RBLOCK])
        tmp11 = _tmp10 + tmp9
        _tmp10 = tl.where(rmask & xmask, tmp11, _tmp10)
```

where everything is fused into a single reduction loop.


I'll have a look into the benchmark cases and see if anything sticks out to me.

**Comment by peterbell10** [comment]
So it turns out that all of these examples are using the exact same function
```python
def create_position_ids_from_input_ids(input_ids, padding_idx):
    mask = input_ids.ne(padding_idx).int()
    incremental_indices = torch.cumsum(mask, dim=1).type_as(mask) * mask
    return incremental_indices.long() + padding_idx
```

Before this PR we generate two kernels and a call to `aten.cumsum`, after this PR we generate a single fused kernel and it is faster.  The issue is that `input_ids` is just really small ~8K elements, so it was never taking that much of the execution time to being with. 

If I compare just the graph break that contains the cumsum in `AllenaiLongformerBase` for example, I see a 1.1x speedup. And that also includes doing various embeddings. However, that graph was already pretty fast compared to the more compute heavy graphs that do lots of matmuls. So the overall speedup is not too significant.

**Review by lezcano — APPROVED** [review]
Thank you for the analysis!

**Comment by jithunnair-amd** [comment]
cc @jataylo @dllehr-amd Please see if we can conditionalize some logic to avoid this issue for rocm

**Comment by huydhn** [comment]
The error shows on the log is https://github.com/pytorch/pytorch/actions/runs/6113041654/job/16592307448#step:11:2555

```
AttributeError: module 'triton.language' has no attribute 'cumprod'
```

So this looks like it's coming from rocm triton?

**Comment by jataylo** [comment]
> The error shows on the log is https://github.com/pytorch/pytorch/actions/runs/6113041654/job/16592307448#step:11:2555
> 
> ```
> AttributeError: module 'triton.language' has no attribute 'cumprod'
> ```
> 
> So this looks like it's coming from rocm triton?

We do not support cumprod/cumsum in our triton branch yet, this is a relatively new development and requires some triton fixes on our side to support and then a triton commit pin.

For now we will need to conditionalise out the lowering from ROCm similar to the way handeled for tl.reduce previously https://github.com/pytorch/pytorch/pull/100089/files

**Comment by Chillee** [comment]
Very cool!

**Review by atalman — APPROVED** [review]

**Comment by ezyang** [comment]
@peterbell10 are you planning on also adding a higher order op for scan :)

**Comment by peterbell10** [comment]
@ezyang yeah I will work on generalizing this as a higher order op.

**Pull request closed without merging**